lib/Sema/SemaBounds.cpp

```diff
@@ -936,6 +934,20 @@ namespace {
         return true;
       }
 
+      if (CK == CK_DynamicPtrBounds) {
+        BoundsExpr *SrcBounds = S.InferRValueBounds(E);
```

Comment:
I am confused by this code.   Why are you inferring the rvalue bounds for the entire cast expression?   We also should not be modifying the bounds expression with E->setBoundsExpr().  

I think it is enough here to just check that the subexpression has bounds and issue an error message if it does not.